### DISCUSSION TITLE
Accessing UserRepresentation in mapper

### DISCUSSION THREAD

**Discussion by japoorv** [keycloak/keycloak#22752] [Q&A] [answered]
I want to determine if a user has logged in via Google/SAML or direct credentials. 
I read that this can be achieved if we know the UserRepresentation of the user and can look under `federatedIdentities`.

The transform/setClaims method all give access to UserSessionModel (which has UserModel) but not UserRepresentation. How do I get that ? 
Do I need to hit the REST API (seems like there should be another way, since all of this is already happening inside keycloak) ?

**Comment by justin-tay** [comment]
You can get the `FederatedIdentityModel` in this way.

```java
@Override
protected void setClaim(final IDToken token,
                        final ProtocolMapperModel mappingModel,
                        final UserSessionModel userSession,
                        final KeycloakSession keycloakSession,
                        final ClientSessionContext clientSessionCtx) {
    UserModel user = userSession.getUser();
    RealmModel realm = userSession.getRealm();
    keycloakSession.users().getFederatedIdentitiesStream(realm, user);
}
```

**Reply by japoorv** [reply]
Ah neat. 
Thanks ! 
Lastly how can I access the attributes being passed by the SAML IDP to keycloak ?

**Reply by justin-tay** [reply]
You can try and see if what you want is available as a user session note in `userSession.getNotes()`.

For instance if you wanted to know the brokered identity provider for that session.

```java
userSession.getNote("identity_provider");
```

For the federated id token when using OIDC

```java
userSession.getNote("FEDERATED_ID_TOKEN");
```

That said for SAML I suspect to access the SAML Assertion you would need to have an Identity Provider Mapper to first map the attributes on the assertion to user attributes on the user model first.

**Comment by japoorv** [comment]
Thanks @justin-tay for the replies. 
I can use the federated identities for getting the login strategies, but what if the user has in the past logged in via different IdPs (say Google, Facebook), in that case the `federatedIdentitiesStream` would have both. 
So essentially what I want is the current federated identity that this is person is trying to log in through, the mapper would get executed in that session, so there should be a way to get the login strategy of the current flow.

**Reply by justin-tay** [reply, marked as answer]
As indicated in my previous reply. If you wanted to know the brokered identity provider for that particular login session it should be indicated in the user session notes.

```java
userSession.getNote("identity_provider");
```

**Reply by japoorv** [reply]
Oh cool, missed it.
Thanks!!
